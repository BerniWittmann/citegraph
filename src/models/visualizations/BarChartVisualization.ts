import Visualization, { VisualizationBaseFields } from './Visualization'

export interface BarChartVisualizationFields {}

export default class BarChartVisualization extends Visualization implements BarChartVisualizationFields {
  static key: string = 'barchart'
  key: string = 'barchart'
  static title: string = 'visualizations.information.bar_chart.title'
  static description: string = 'visualizations.information.bar_chart.description'
  static longDescription: string = 'visualizations.information.bar_chart.long_description'
  static imageUrl: string = 'visualization_images/bar_chart.png'

  constructor ({ id, name, progress }: VisualizationBaseFields & BarChartVisualizationFields) {
    super({ id, name, progress })
  }
}
