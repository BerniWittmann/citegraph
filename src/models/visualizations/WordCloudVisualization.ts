import Visualization, { VisualizationBaseFields } from './Visualization'

export interface WordCloudVisualizationParameters {
  pruneLeastCitedPercentage: number
}

export interface WordCloudDataPoint {
  word: string
  weight: number
}
export interface WordCloudData {
  data: Array<WordCloudDataPoint>
}

export interface WordCloudVisualizationFields {
  data?: WordCloudData
}

export default class WordCloudVisualization extends Visualization implements WordCloudVisualizationFields {
  static key: string = 'wordcloud'
  key: string = 'wordcloud'
  static title: string = 'visualizations.information.word_cloud.title'
  static description: string = 'visualizations.information.word_cloud.description'
  static longDescription: string = 'visualizations.information.word_cloud.long_description'
  static imageUrl: string = 'visualization_images/word_cloud.jpg'
  parameters: WordCloudVisualizationParameters
  data?: WordCloudData

  constructor ({ id, name, progress, data }: VisualizationBaseFields & WordCloudVisualizationFields) {
    super({ id, name, progress })
    this.parameters = {
      pruneLeastCitedPercentage: 5
    }
    this.data = data
  }
}
