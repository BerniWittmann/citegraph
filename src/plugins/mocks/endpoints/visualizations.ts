import { AxiosRequestConfig } from 'axios'
import MockAdapter from 'axios-mock-adapter'
import db from '../db'
import Project from '@/models/project'
import Visualization from '@/models/visualizations/Visualization'

const projects = db.getCollection('projects')
const visualizations = db.getCollection('visualizations')

function getProjectId (config: AxiosRequestConfig): number {
  const urlParts: Array<string> = config.url!.split('/')
  return parseInt(urlParts[4])
}

function getVisualizationId (config: AxiosRequestConfig): string {
  const urlParts: Array<string> = config.url!.split('/')
  return urlParts[6]
}

export function setupMocks (mock: MockAdapter) {
  mock.onGet(/\/projects\/\d+\/visualizations$/).reply((config: AxiosRequestConfig) => {
    const project: Resultset<Project> = projects.findOne({ 'id': getProjectId(config) })
    if (!project) {
      return [404]
    }
    return [200, visualizations.find()]
  })

  mock.onGet(/\/projects\/\d+\/visualizations\/\d+$/).reply((config: AxiosRequestConfig) => {
    const project: Resultset<Project> = projects.findOne({ 'id': getProjectId(config) })
    if (!project) {
      return [404]
    }
    const visualization: Resultset<Visualization> = visualizations.findOne({ 'id': getVisualizationId(config) })
    if (!visualization) {
      return [404]
    }
    return [200, visualization]
  })
}
