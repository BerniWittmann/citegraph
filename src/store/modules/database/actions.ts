import { ActionTree } from 'vuex'
import { RootState } from '@/store/types'
import { DatabaseState } from '@/store/modules/database/types'
import Axios from 'axios'

export const actions: ActionTree<DatabaseState, RootState> = {
  async clearDatabase ({ dispatch }): Promise<undefined> {
    try {
      await Axios(`/database/clear`, {
        method: 'POST'
      })
      dispatch('toasts/showSuccess', 'settings.database.clear.success', { root: true })
      return undefined
    } catch (error) {
      console.error(error)
      dispatch('toasts/showError', 'settings.database.clear.error', { root: true })
      return error
    }
  },

  async reseedDatabase ({ dispatch }): Promise<undefined> {
    try {
      await Axios(`/database/reseed`, {
        method: 'POST'
      })
      dispatch('toasts/showSuccess', 'settings.database.reseed.success', { root: true })
      return undefined
    } catch (error) {
      console.error(error)
      dispatch('toasts/showError', 'settings.database.reseed.error', { root: true })
      return error
    }
  }
}
