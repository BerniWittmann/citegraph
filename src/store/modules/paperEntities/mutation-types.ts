export const SET_ENTITIES = 'SET_ENTITIES'
export const SET_ENTITY_TYPE = 'SET_ENTITY_TYPE'
export const SET_ENTITY_COUNT = 'SET_ENTITY_COUNT'
