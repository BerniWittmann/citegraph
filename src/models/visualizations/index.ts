import WordCloudVisualization, { WordCloudVisualizationFields } from './WordCloudVisualization'
import { VisualizationBaseFields } from '@/models/visualizations/Visualization'
import BarChartVisualization, { BarChartVisualizationFields } from '@/models/visualizations/BarChartVisualization'

export const visualizations = [WordCloudVisualization, BarChartVisualization]

const visualizationsKeyMap: Record<string, any> = {}
visualizations.forEach((visualization) => {
  visualizationsKeyMap[visualization.key] = visualization
})

export {
  visualizationsKeyMap
}

export type VisalizationFields = VisualizationBaseFields & (
  WordCloudVisualizationFields &
  BarChartVisualizationFields
)
