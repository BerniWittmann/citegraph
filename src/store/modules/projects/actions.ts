import Axios from 'axios'
import { ActionTree } from 'vuex'
import { ProjectsState } from '@/store/modules/projects/types'
import { RootState } from '@/store/types'
import Project from '@/models/project'
import * as types from './mutation-types'

export const actions: ActionTree<ProjectsState, RootState> = {
  async fetchProject ({ commit }, id: string): Promise<undefined> {
    try {
      const response = await Axios(`/projects/${id}`)
      const payload: Project = new Project(response && response.data)
      commit(types.SET_ACTIVE_PROJECT, payload)
      commit(types.OPEN_PROJECT, payload)
      return undefined
    } catch (error) {
      console.error(error)
      commit(types.UNSET_ACTIVE_PROJECT)
      return error
    }
  },

  unsetActiveProject ({ commit }): void {
    commit(types.UNSET_ACTIVE_PROJECT)
  },

  async fetchProjects ({ commit }): Promise<undefined> {
    try {
      const response = await Axios(`/projects`)
      const projects = response.data.map((project: Project) => new Project(project))
      commit(types.SET_PROJECTS, projects)
      return undefined
    } catch (error) {
      console.error(error)
      return error
    }
  },

  openProject ({ commit }, project: Project): void {
    commit(types.OPEN_PROJECT, project)
  },

  closeProject ({ commit }, project: Project): void {
    commit(types.CLOSE_PROJECT, project)
  },

  async createProject ({ commit }, data: Project): Promise<Project | undefined> {
    try {
      const response = await Axios(`/projects`, {
        method: 'POST',
        data: data
      })
      const project = new Project(response.data)
      commit(types.ADD_PROJECT, project)
      return project
    } catch (error) {
      console.error(error)
      return error
    }
  }
}
