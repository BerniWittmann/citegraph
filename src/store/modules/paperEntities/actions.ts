import { ActionTree } from 'vuex'
import { RootState } from '@/store/types'
import { PaperEntitiesState } from './types'
import { PaperEntityQueryParameters } from '@/models/paperEntities/query'
import Axios from 'axios'
import * as types from '@/store/modules/paperEntities/mutation-types'
import jsLogger from 'js-logger'
import { entityKeysMap, PaperEntity, PaperEntityFields } from '@/models/paperEntities'
import { PaperEntityMutationParameters } from '@/models/paperEntities/mutation'

const logger = jsLogger.get('paperEntities/actions')

export const actions: ActionTree<PaperEntitiesState, RootState> = {
  async fetchEntities ({ commit, dispatch }, params: PaperEntityQueryParameters): Promise<undefined> {
    try {
      const EntityClass = entityKeysMap[params.entityType]
      const response = await Axios(`/projects/${params.projectId}/paper-entities`, {
        method: 'POST',
        data: {
          query: EntityClass.getQuery(params)
        }
      })
      const data = response.data.data[EntityClass.queryName]
      const totalCount = data.count
      const payload: Array<PaperEntity> = EntityClass.transformer.fetchCollection(data[EntityClass.schemaName])
      commit(types.SET_ENTITIES, payload)
      commit(types.SET_ENTITY_COUNT, totalCount)
      commit(types.SET_ENTITY_TYPE, EntityClass.key)
      return undefined
    } catch (error) {
      logger.error(error)
      dispatch('toasts/showError', 'project.explore.fetch_error', { root: true })
      commit(types.SET_ENTITIES, [])
      commit(types.SET_ENTITY_COUNT, -1)
      commit(types.SET_ENTITY_TYPE, undefined)
      return error
    }
  },

  async fetchEntity ({ commit, dispatch }, params: PaperEntityQueryParameters): Promise<void> {
    try {
      const EntityClass = entityKeysMap[params.entityType]
      const response = await Axios(`/projects/${params.projectId}/paper-entities`, {
        method: 'POST',
        data: {
          query: EntityClass.getQuery(params)
        }
      })
      const data = response.data.data[EntityClass.queryName]
      const payload: PaperEntity = EntityClass.transformer.fetch(data[EntityClass.schemaName])
      commit(types.SET_ACTIVE_ENTITY, payload)
      return undefined
    } catch (error) {
      logger.error(error)
      dispatch('toasts/showError', 'project.explore.fetch_single_error', { root: true })
      commit(types.SET_ACTIVE_ENTITY, undefined)
      return error
    }
  },

  async updateEntity ({ commit, dispatch }, {
    params,
    data
  }: { params: PaperEntityMutationParameters, data: Partial<PaperEntityFields>}): Promise<void> {
    try {
      const EntityClass = entityKeysMap[params.entityType]
      const response = await Axios(`/projects/${params.projectId}/paper-entities`, {
        method: 'POST',
        data: {
          mutation: EntityClass.getMutation(params, EntityClass.transformer.send(data))
        }
      })
      const responseData = response.data.data[EntityClass.queryName]
      const payload: PaperEntity = EntityClass.transformer.fetch(responseData)
      commit(types.UPDATE_ENTITY, payload)
      return undefined
    } catch (error) {
      logger.error(error)
      dispatch('toasts/showError', 'project.explore.update_single_error', { root: true })
      return error
    }
  }
}
