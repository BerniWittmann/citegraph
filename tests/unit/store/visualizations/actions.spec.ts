import moxios from 'moxios'
import { actions } from '@/store/modules/visualizations/actions'
import * as mutationTypes from '@/store/modules/visualizations/mutation-types'
import WordCloudVisualization from '@/models/visualizations/WordCloudVisualization'
import BarChartVisualization from '@/models/visualizations/BarChartVisualization'
import WordCloudVisualizationTransformer from '@/transformers/WordCloudVisualizationTransformer'

describe('store/modules/visualizations/actions', () => {
  it('provides the actions', () => {
    expect(actions).toMatchSnapshot()
  })

  describe('fetchVisualizations', () => {
    const visualizationsData = [
      { id: '42', name: 'My undefined Project' },
      { id: '43', name: 'My wrong Project', key: 'wrong' },
      { id: '1', name: 'First Project', progress: 0.8, key: WordCloudVisualization.key },
      { id: '2', name: 'Second Project', key: BarChartVisualization.key }
    ]

    beforeEach(() => {
      moxios.install()
    })
    afterEach(() => {
      moxios.uninstall()
    })
    it('fetches all projects', (done) => {
      moxios.stubRequest('/projects/42/visualizations', {
        status: 200,
        response: visualizationsData
      })

      const onFulfilled = jest.fn()
      const commit = jest.fn()
      const action = actions.fetchVisualizations as Function

      action({ commit }, 42).then(onFulfilled)

      moxios.wait(() => {
        expect(onFulfilled).toHaveBeenCalled()
        expect(commit).toHaveBeenCalledWith(mutationTypes.SET_VISUALIZATIONS, [
          new WordCloudVisualization({ id: '1', name: 'First Project', progress: 0.8 }),
          new BarChartVisualization({ id: '2', name: 'Second Project' })
        ])
        done()
      })
    })

    it('handles the error when one appeared during the fetching', (done) => {
      moxios.stubRequest('/projects/42/visualizations', {
        status: 500,
        response: 'An error occured'
      })

      const onFulfilled = jest.fn()
      const commit = jest.fn()
      const dispatch = jest.fn()
      const action = actions.fetchVisualizations as Function

      action({ commit, dispatch }, 42).then(onFulfilled)

      moxios.wait(() => {
        expect(onFulfilled).toHaveBeenCalledWith(new Error('Request failed with status code 500'))
        expect(commit).not.toHaveBeenCalled()
        expect(dispatch).toHaveBeenCalledWith('toasts/showError', 'visualizations.fetch_error', { root: true })
        done()
      })
    })
  })

  describe('unsetCurrentVisualization', () => {
    it('unsets the visualization', () => {
      const commit = jest.fn()
      const action = actions.unsetCurrentVisualization as Function

      action({ commit })
      expect(commit).toHaveBeenCalledWith(mutationTypes.SET_CURRENT_VISUALIZATION, undefined)
    })
  })

  describe('fetchVisualization', () => {
    const visualizationData = { id: '1', name: 'First Project', progress: 0.8, key: WordCloudVisualization.key }

    beforeEach(() => {
      moxios.install()
    })
    afterEach(() => {
      moxios.uninstall()
    })
    it('fetches the visualization', (done) => {
      moxios.stubRequest('/projects/42/visualizations/1', {
        status: 200,
        response: visualizationData
      })

      const onFulfilled = jest.fn()
      const commit = jest.fn()
      const action = actions.fetchVisualization as Function

      action({ commit }, {
        projectId: 42,
        visualizationId: 1
      }).then(onFulfilled)

      moxios.wait(() => {
        expect(onFulfilled).toHaveBeenCalled()
        expect(commit).toHaveBeenCalledWith(mutationTypes.SET_CURRENT_VISUALIZATION,
          new WordCloudVisualization({ id: '1', name: 'First Project', progress: 0.8 })
        )
        done()
      })
    })

    it('handles the error when one appeared during the fetching', (done) => {
      moxios.stubRequest('/projects/42/visualizations/1', {
        status: 500,
        response: 'An error occured'
      })

      const onFulfilled = jest.fn()
      const commit = jest.fn()
      const dispatch = jest.fn()
      const action = actions.fetchVisualization as Function

      action({ commit, dispatch }, {
        projectId: 42,
        visualizationId: 1
      }).then(onFulfilled)

      moxios.wait(() => {
        expect(onFulfilled).toHaveBeenCalledWith(new Error('Request failed with status code 500'))
        expect(commit).not.toHaveBeenCalled()
        expect(dispatch).toHaveBeenCalledWith('toasts/showError', 'visualizations.fetch_single_error', { root: true })
        done()
      })
    })

    it('handles the error when one appeared during the parsing', (done) => {
      moxios.stubRequest('/projects/42/visualizations/1', {
        status: 200,
        response: {
          ...visualizationData,
          key: 'invalid'
        }
      })

      const onFulfilled = jest.fn()
      const commit = jest.fn()
      const dispatch = jest.fn()
      const action = actions.fetchVisualization as Function

      action({ commit, dispatch }, {
        projectId: 42,
        visualizationId: 1
      }).then(onFulfilled)

      moxios.wait(() => {
        expect(onFulfilled).toHaveBeenCalledWith(new Error('Could not parse visualization'))
        expect(commit).not.toHaveBeenCalled()
        expect(dispatch).toHaveBeenCalledWith('toasts/showError', 'visualizations.fetch_single_error', { root: true })
        done()
      })
    })
  })

  describe('createVisualization', () => {
    const visualizationData = { id: '1', name: 'First Project', progress: 0.8, key: WordCloudVisualization.key }

    beforeEach(() => {
      moxios.install()
    })
    afterEach(() => {
      moxios.uninstall()
    })
    it('creates the visualization', (done) => {
      moxios.stubRequest('/projects/42/visualizations', {
        status: 200,
        response: WordCloudVisualizationTransformer.send(visualizationData as any)
      })

      const onFulfilled = jest.fn()
      const commit = jest.fn()
      const dispatch = jest.fn()
      const action = actions.createVisualization as Function

      action({ commit, dispatch }, {
        projectId: 42,
        visualizationData,
        visualizationClass: WordCloudVisualization
      }).then(onFulfilled)

      moxios.wait(() => {
        expect(onFulfilled).toHaveBeenCalled()
        expect(commit).toHaveBeenCalledWith(mutationTypes.ADD_VISUALIZATION,
          new WordCloudVisualization({ id: '1', name: 'First Project', progress: 0.8 })
        )
        expect(dispatch).toHaveBeenCalledWith('toasts/showSuccess', 'visualizations.add.successful', { root: true })
        done()
      })
    })

    it('handles the error when one appeared during the creation', (done) => {
      moxios.stubRequest('/projects/42/visualizations', {
        status: 500,
        response: 'An error occured'
      })

      const onFulfilled = jest.fn()
      const commit = jest.fn()
      const dispatch = jest.fn()
      const action = actions.createVisualization as Function

      action({ commit, dispatch }, {
        projectId: 42,
        visualizationData,
        visualizationClass: WordCloudVisualization
      }).then(onFulfilled)

      moxios.wait(() => {
        expect(onFulfilled).toHaveBeenCalledWith(new Error('Request failed with status code 500'))
        expect(commit).not.toHaveBeenCalled()
        expect(dispatch).toHaveBeenCalledWith('toasts/showError', 'visualizations.add.error', { root: true })
        done()
      })
    })

    it('handles the error when one appeared during the parsing', (done) => {
      moxios.stubRequest('/projects/42/visualizations', {
        status: 200,
        response: {
          ...WordCloudVisualizationTransformer.send(visualizationData as any),
          key: 'invalid'
        }
      })

      const onFulfilled = jest.fn()
      const commit = jest.fn()
      const dispatch = jest.fn()
      const action = actions.createVisualization as Function

      action({ commit, dispatch }, {
        projectId: 42,
        visualizationData,
        visualizationClass: WordCloudVisualization
      }).then(onFulfilled)

      moxios.wait(() => {
        expect(onFulfilled).toHaveBeenCalledWith(new Error('Could not parse visualization'))
        expect(commit).not.toHaveBeenCalled()
        expect(dispatch).toHaveBeenCalledWith('toasts/showError', 'visualizations.add.error', { root: true })
        done()
      })
    })
  })

  describe('updateVisualization', () => {
    const visualizationData = { id: '1', name: 'First Project', progress: 0.8, key: WordCloudVisualization.key }

    beforeEach(() => {
      moxios.install()
    })
    afterEach(() => {
      moxios.uninstall()
    })
    it('updates the visualization', (done) => {
      moxios.stubRequest('/projects/42/visualizations/1', {
        status: 200,
        response: WordCloudVisualizationTransformer.send(visualizationData as any)
      })

      const onFulfilled = jest.fn()
      const commit = jest.fn()
      const dispatch = jest.fn()
      const action = actions.updateVisualization as Function

      action({ commit, dispatch }, {
        projectId: 42,
        visualizationData,
        visualizationClass: WordCloudVisualization
      }).then(onFulfilled)

      moxios.wait(() => {
        expect(onFulfilled).toHaveBeenCalled()
        expect(commit).toHaveBeenCalledWith(mutationTypes.UPDATE_VISUALIZATION,
          new WordCloudVisualization({ id: '1', name: 'First Project', progress: 0.8 })
        )
        expect(dispatch).toHaveBeenCalledWith('toasts/showSuccess', 'visualization.edit.successful', { root: true })
        done()
      })
    })

    it('handles the error when one appeared during the updating', (done) => {
      moxios.stubRequest('/projects/42/visualizations/1', {
        status: 500,
        response: 'An error occured'
      })

      const onFulfilled = jest.fn()
      const commit = jest.fn()
      const dispatch = jest.fn()
      const action = actions.updateVisualization as Function

      action({ commit, dispatch }, {
        projectId: 42,
        visualizationData,
        visualizationClass: WordCloudVisualization
      }).then(onFulfilled)

      moxios.wait(() => {
        expect(onFulfilled).toHaveBeenCalledWith(new Error('Request failed with status code 500'))
        expect(commit).not.toHaveBeenCalled()
        expect(dispatch).toHaveBeenCalledWith('toasts/showError', 'visualization.edit.error', { root: true })
        done()
      })
    })

    it('handles the error when one appeared during the parsing', (done) => {
      moxios.stubRequest('/projects/42/visualizations/1', {
        status: 200,
        response: {
          ...WordCloudVisualizationTransformer.send(visualizationData as any),
          key: 'invalid'
        }
      })

      const onFulfilled = jest.fn()
      const commit = jest.fn()
      const dispatch = jest.fn()
      const action = actions.updateVisualization as Function

      action({ commit, dispatch }, {
        projectId: 42,
        visualizationData,
        visualizationClass: WordCloudVisualization
      }).then(onFulfilled)

      moxios.wait(() => {
        expect(onFulfilled).toHaveBeenCalledWith(new Error('Could not parse visualization'))
        expect(commit).not.toHaveBeenCalled()
        expect(dispatch).toHaveBeenCalledWith('toasts/showError', 'visualization.edit.error', { root: true })
        done()
      })
    })
  })
})
