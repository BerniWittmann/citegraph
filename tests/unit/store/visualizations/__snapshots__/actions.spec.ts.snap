// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`store/modules/visualizations/actions provides the actions 1`] = `
Object {
  "createVisualization": [Function],
  "fetchVisualization": [Function],
  "fetchVisualizations": [Function],
  "unsetCurrentVisualization": [Function],
  "updateVisualization": [Function],
}
`;
